server/storage/database_test.go

```diff
@@ -2,290 +2,326 @@ package storage
 	"testing"
 
-	"github.com/DATA-DOG/go-sqlmock"
-	"github.com/go-sql-driver/mysql"
+	"github.com/endophage/gotuf/data"
+	"github.com/jinzhu/gorm"
+	_ "github.com/mattn/go-sqlite3"
 	"github.com/stretchr/testify/assert"
 )
 
-func TestMySQLUpdateCurrent(t *testing.T) {
-	db, err := sqlmock.New()
-	assert.Nil(t, err, "Could not initialize mock DB")
-	s := NewMySQLStorage(db)
-	update := MetaUpdate{
-		Role:    "root",
-		Version: 0,
-		Data:    []byte("1"),
-	}
-	sqlmock.ExpectQuery("SELECT count\\(\\*\\) FROM `tuf_files` WHERE `gun`=\\? AND `role`=\\? AND `version`>=\\?\\;").WithArgs(
-		"testGUN",
-		update.Role,
-		update.Version,
-	).WillReturnRows(sqlmock.RowsFromCSVString([]string{"count(*)"}, "0"))
-	sqlmock.ExpectExec("INSERT INTO `tuf_files` \\(`gun`, `role`, `version`, `data`\\) VALUES \\(\\?,\\?,\\?,\\?\\);").WithArgs(
-		"testGUN",
-		update.Role,
-		update.Version,
-		update.Data,
-	).WillReturnResult(sqlmock.NewResult(0, 1))
-
-	err = s.UpdateCurrent(
-		"testGUN",
-		update,
-	)
-	assert.Nil(t, err, "UpdateCurrent errored unexpectedly: %v", err)
-
-	// There's a bug in the mock lib
-	//err = db.Close()
-	//assert.Nil(t, err, "Expectation not met: %v", err)
+// GormTUFFile represents a TUF file in the database
+type GormTUFFile struct {
+	ID      int    `sql:"AUTO_INCREMENT" gorm:"primary_key"`
+	Gun     string `sql:"type:varchar(255);not null"`
+	Role    string `sql:"type:varchar(255);not null"`
+	Version int
+	Data    []byte `sql:"type:longblob"`
 }
 
-func TestMySQLUpdateCurrentError(t *testing.T) {
-	db, err := sqlmock.New()
-	assert.Nil(t, err, "Could not initialize mock DB")
-	s := NewMySQLStorage(db)
-	update := MetaUpdate{
-		Role:    "root",
-		Version: 0,
-		Data:    []byte("1"),
-	}
-	sqlmock.ExpectQuery("SELECT count\\(\\*\\) FROM `tuf_files` WHERE `gun`=\\? AND `role`=\\? AND `version`>=\\?\\;").WithArgs(
-		"testGUN",
-		update.Role,
-		update.Version,
-	).WillReturnRows(sqlmock.RowsFromCSVString([]string{"count(*)"}, "0"))
-	sqlmock.ExpectExec("INSERT INTO `tuf_files` \\(`gun`, `role`, `version`, `data`\\) VALUES \\(\\?,\\?,\\?,\\?\\);").WithArgs(
-		"testGUN",
-		update.Role,
-		update.Version,
-		update.Data,
-	).WillReturnError(
-		&mysql.MySQLError{
-			Number:  1022,
-			Message: "Duplicate key error",
-		},
-	)
-
-	err = s.UpdateCurrent(
-		"testGUN",
-		update,
-	)
-	assert.NotNil(t, err, "Error should not be nil")
-	assert.IsType(t, &ErrOldVersion{}, err, "Expected ErrOldVersion error type, got: %v", err)
-
-	// There's a bug in the mock lib
-	//err = db.Close()
-	//assert.Nil(t, err, "Expectation not met: %v", err)
+// TableName sets a specific table name for GormTUFFile
+func (g GormTUFFile) TableName() string {
+	return "tuf_files"
 }
 
-func TestMySQLUpdateMany(t *testing.T) {
-	db, err := sqlmock.New()
-	assert.Nil(t, err, "Could not initialize mock DB")
-	s := NewMySQLStorage(db)
-	update1 := MetaUpdate{
+// GormTimestampKey represents a single timestamp key in the database
+type GormTimestampKey struct {
```

Comment:
You're absolutely correct, uniqueness was being enforced by the fact `Gun` was the primary key. We will need to add a separate uniqueness constraint.

Because there are no foreign key relationships, we can do this migration of primary key. Sadly we won't be able to do it in Gorm, it will have to be a hand crafted sql file as it's a little fiddly. You can take a crack at it and add it to this PR or I can send you a PR against this branch on your fork. Let me know.

Once we merge this PR, I'll look at updating the DB. I believe the changes proposed will be compatible with the existing production code but I'll take another look when the time comes.